File: Erlang/racer_proxy.erl
%% to run the program:
%%   erl -compile racer_proxy.erl && erl -p -run racer_proxy start -run init stop -noshell
-module(racer_proxy).

-export([main/1, proxy/2, start/0]).

start() -> main(1).

%% the worker
worker_main(From, Arg, {}) ->
    % wait one second (emulate work)
    timer:sleep(timer:seconds(1)),
    % send the answer back
    From ! { worker_result, Arg }.

%% the proxy hub
loop(_, NumWorkers, DeadWorkers) when NumWorkers == DeadWorkers ->
    exit(all_died);
loop(From, NumWorkers, DeadWorkers) ->
    receive
        { 'EXIT', _, normal } ->
            % worker went down normally (we should have received a result)
            loop(From, NumWorkers, DeadWorkers);
        { 'EXIT', _, _ } ->
            % worker died, need to account for that
            loop(From, NumWorkers, DeadWorkers + 1);
        { worker_result, Res } ->
            % got a worker result, communicate back
            From ! { proxy_done, self(), Res },
            exit(ok);
        _ -> exit(unexpected)
    end.

proxy(From, Request) ->
    % we want to get notified when the workers die, even though we'll spawn with
    % link
    process_flag(trap_exit, true),
    NumWorkers = 10000,
    Me = self(),
    lists:map(fun(WorkerId) ->
                      spawn_link(fun() -> worker_main(Me, WorkerId, Request) end)
              end,
              lists:seq(1, NumWorkers)),
    loop(From, NumWorkers, 0).

%% the main program
main(_) ->
    process_flag(trap_exit, true),
    Req = {},
    _ = spawn_link(racer_proxy, proxy, [self(), Req]),
    receive
        {'EXIT', Pid, Reason} ->
            io:format("Exiting because ~p died: (~p) ~n", [Pid, Reason]);
        {proxy_done, _, Res} ->
            io:format("Got result: ~p~n", [Res]);
        Unexpected ->
            io:format("ERROR: unexpected stuff: ~p~n", [Unexpected]),
            exit(error)
    end.

File: MockSwift-Erlang-Actors/racer_proxy.swift
/* made up syntax:
  1. the `spawn` keyword to spawn an actor
     with it's `spawn linked` variant to spawn a linked actor (deaths propagate
     both ways)

  2. func foo(arg: Arg) actor<Message>

     creates an actor with argument `arg` of type `Arg` which can receive messages
     of type `Message`. Actors are always throwing. Errors get delivered to
     as messages
*/

typealias Request = ()
typealias Result = Int

enum ActorExit {
    case normal
    case error(Error)
}

protocol ActorMessage {
    init(exited: ActorExit)
}

class Actor<MessageType>: ActorMessage {
    ...
}

enum ProxyHubIn: ActorMessage {
    case workerResult(Result)
    case workerDied(ActorExit)

    init(exited: ActorExit) {
        self = .workerDied(exited)
    }
}

enum MainActorIn: ActorMessage {
    case proxyDone(Actor<>, Result)
    case workerDied(ActorExit)

    init(exited: ActorExit) {
        self = .workerDied(exited)
    }
}

struct AllDiedError: Error {}

// the worker
func workerMain(from: Actor<ProxyHubIn>, workerId: Int, request: ()) -> Void {
    // wait one second (emulate work)
    Actor.sleep(1)
    from.send(.workerResult(workerId))
}

// the proxy hub
func proxy(from: Actor<MainActorIn>, request: Request) actor<ProxyHubIn> -> Void {
    // we want to get notified when the workers die, even though we'll spawn with
    // link
    Actor.processFlag(.trapExit, true)
    let numWorkers = 10000
    let me: actor<ProxyHubIn> = Actor.self
    for wid in 1...numWorkers {
        _ = spawn linked workerMain(from: me, workerId: wid, request: ())
    }
    var deadWorkers = 0
    while true {
        if numWorkers == deadWorkers {
            throw AllDiedError()
        }
        switch Actor.receive() {
            case .workerDied(.normal):
                // worker went down normally (we should have received a result)
                continue
            case .workerDied(.error(_)):
                // worker died, need to account for that
                deadWorkers += 1
                continue
            case .workerResult(let res):
                // got a worker result, communicate back
                from.send(.proxyDone(Actor.self, res))
                break
        }
    }
}

func main() actor<MainActorIn> {
    Actor.processFlag(.trapExit, true)
    let req = ()
    _ = spawn linked proxy(from: self(), request: req)
    loop: while true {
        let msg = Actor.receive()
        switch msg {
            case .workerDied(.normal):
                continue // ignore, will get a .proxyDone message
            case .workerDied(let reason):
                print("Exiting because the proxy hub died: \(reason)")
                exit(1)
            case .proxyDone(_, let res):
                print("Got result: \(res)")
                exit(0)
        }
    }
}
